src/abe/dippe.c

```diff
@@ -0,0 +1,613 @@
+    // rounds
+    int r = (in_buf_size / MODBYTES_256_56) + ((in_buf_size % MODBYTES_256_56) != 0);
+    // hash buffer
+    char buf[MODBYTES_256_56];
+
+    octet tmp;
+    tmp.val = (char*)&buf;
+    tmp.len = MODBYTES_256_56;
+    tmp.max = MODBYTES_256_56;
+
+    ECP2_BN254_inf(out);
+    ECP2_BN254 tmp_g2;
+
+    for (int i=0; i<r; i++) {
+        for (int j=0; j<MODBYTES_256_56; j++) {
+            if (((i*MODBYTES_256_56)+j) < in_buf_size) buf[j] = in_buf[(i*MODBYTES_256_56)+j]; else buf[j] = '\0';
+            if ((((i*MODBYTES_256_56)+j+1) % MODBYTES_256_56) == 0) {
+                ECP2_BN254_mapit(&tmp_g2, &tmp);
+                ECP2_BN254_add(out, &tmp_g2);
+            }
+        }
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+/**
+ * Clears the scheme and frees allocated memory
+ *
+ * @param dippe A pointer to a cfe_dippe struct
+ */
+void cfe_dippe_free(cfe_dippe *dippe) {
+    mpz_clear(dippe->p);
+    free(dippe->g1_A);
+    free(dippe->g1_UA);
+}
+
+/**
+ * Clears the public key and frees allocated memory
+ *
+ * @param pk A pointer to a cfe_dippe_pub_key struct
+ */
+void cfe_dippe_pub_key_free(cfe_dippe_pub_key *pk) {
+	free(pk->gt_alpha_A);
+	free(pk->g1_W_A);
+}
+
+/**
+ * Clears the secret key and frees allocated memory
+ *
+ * @param sk A pointer to a cfe_dippe_sec_key struct
+ */
+void cfe_dippe_sec_key_free(cfe_dippe_sec_key *sk) {
+	mpz_clear(sk->sigma);
+	cfe_vec_free(&(sk->alpha));
+	cfe_vec_free(&(sk->W));
+}
+
+/**
+ * Clears the ciphertext and frees allocated memory
+ *
+ * @param cipher A pointer to a cfe_dippe_cipher struct
+ */
+void cfe_dippe_cipher_free(cfe_dippe_cipher *cipher) {
+    free(cipher->C0);
+    free(cipher->Ci);
+}
+
+/**
+ * Clears the user secret key and frees allocated memory
+ *
+ * @param usk A pointer to a cfe_dippe_user_secret_key struct
+ */
+void cfe_dippe_user_secret_key_free(cfe_dippe_user_secret_key *usk) {
+    free(usk->Ki);
+}
+
+/**
+ * Initializes the DIPPE scheme and generates the global parameters
+ *
+ * @param dippe A pointer to a cfe_dippe struct
+ * @param assump_size The size of the underlying assumption
+ */
+void cfe_dippe_setup(cfe_dippe *dippe, unsigned int assump_size) {
+
+    // temp vars
+    BIG_256_56 tmp_big;
+    mpz_t tmp_mpz;
+    mpz_init(tmp_mpz);
+
+    dippe->assump_size = assump_size;
+    mpz_init(dippe->p);
+    mpz_from_BIG_256_56(dippe->p, (int64_t*)CURVE_Order_BN254);
+
+    // A (k+1 x k)
+    cfe_vec A;
+    cfe_vec_init(&A, ((assump_size+1) * assump_size));
+    cfe_uniform_sample_vec(&A, dippe->p);
+
```

Comment:
You choose A to be a vector, but instead you could choose to be a matrix with `cfe_mat`. This way you could use already implemented functions as `cfe_mat_transpose` and `cfe_mat_mul` to make things more elegant.

Reply:
You're right, I changed the code to utilize the already existent matrix struct.